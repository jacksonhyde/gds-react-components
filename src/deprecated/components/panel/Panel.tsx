import React, { HTMLProps, isValidElement } from 'react';
import classNames from 'classnames';
import { Row, Col } from '../../../components/layout';
import PanelContext, { PanelContextType } from './PanelContext';
import useDevWarning from '../../../util/hooks/UseDevWarning';
import { PanelDeprecationWarning } from '../../warnings';

interface PanelProps extends HTMLProps<HTMLDivElement> {
  grey?: boolean;
  label?: string;
  labelProps?: HTMLProps<HTMLHeadingElement>;
}

interface Panel extends React.FC<PanelProps> {
  Group: React.FC<HTMLProps<HTMLDivElement>>;
}

const BasePanel: React.FC<PanelProps> = ({
  label,
  grey,
  className,
  labelProps,
  children,
  ...rest
}) => {
  useDevWarning(PanelDeprecationWarning);

  return (
    <div
      className={classNames(
        { 'govuk-panel': !label },
        { 'govuk-panel--grey': grey },
        { 'govuk-panel-with-label': label },
        className,
      )}
      {...rest}
    >
      {label ? (
        <h3 className="govuk-panel-with-label__label" {...labelProps}>
          {label}
        </h3>
      ) : null}
      {children}
    </div>
  );
};

const Panel: Panel = (props) => {
  const PanelGroupContext = React.useContext<PanelContextType | null>(PanelContext);
  let panelWidth: 'one-half' | 'one-third' | 'one-quarter' | 'full';
  if (PanelGroupContext !== null) {
    const { panelCount } = PanelGroupContext;
    switch (panelCount) {
      case 2:
        panelWidth = 'one-half';
        break;
      case 3:
        panelWidth = 'one-third';
        break;
      case 4:
        panelWidth = 'one-quarter';
        break;
      default:
        panelWidth = 'full';
        break;
    }
    return (
      <Col width={panelWidth} className="govuk-panel-group__item">
        <BasePanel {...props} />
      </Col>
    );
  }
  return <BasePanel {...props} />;
};

const PanelGroup: React.FC<HTMLProps<HTMLDivElement>> = ({ className, children, ...rest }) => {
  let panelCount = 0;
  React.Children.forEach(children, (child) => {
    if (child && isValidElement(child) && child.type === Panel) {
      panelCount += 1;
    }
  });
  return (
    <Row className={classNames('govuk-panel-group', className)} {...rest}>
      <PanelContext.Provider value={{ panelCount }}>{children}</PanelContext.Provider>
    </Row>
  );
};

Panel.Group = PanelGroup;

export default Panel;
