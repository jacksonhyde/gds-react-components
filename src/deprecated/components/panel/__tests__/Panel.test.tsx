import React from 'react';
import { mount } from 'enzyme';
import Panel from '..';
import { PanelDeprecationWarning } from '../../../warnings';

jest.spyOn(console, 'warn').mockImplementation();

describe('Panel', () => {
  it('prints console deprecation warning', () => {
    const element = mount(<Panel />);

    // eslint-disable-next-line no-console
    expect(console.warn).toHaveBeenCalled();
    // eslint-disable-next-line no-console
    expect((console.warn as jest.Mock).mock.calls[0][0]).toBe(PanelDeprecationWarning);

    element.unmount();
  });

  it('matches snapshot', () => {
    const element = mount(<Panel />);
    expect(element).toMatchSnapshot('Panel');
    element.unmount();
  });

  it('adds correct classes when grey', () => {
    const element = mount(<Panel grey />);
    const renderedElement = element.render();
    expect(renderedElement.hasClass('govuk-panel--grey')).toBeTruthy();
    element.unmount();
  });

  it('adds correct attributes when a label is added', () => {
    const element = mount(<Panel label="Label" />);
    const renderedElement = element.render();
    expect(renderedElement.hasClass('govuk-panel-with-label')).toBeTruthy();
    expect(renderedElement.find('.govuk-panel-with-label__label').text()).toBe('Label');
    element.unmount();
  });

  describe('Panel.Group', () => {
    it('assigns correct widths', () => {
      const twoColumns = mount(
        <Panel.Group>
          <Panel />
          <Panel />
        </Panel.Group>,
      );
      expect(twoColumns.find('.govuk-grid-column-one-half').length).toBe(2);
      twoColumns.unmount();

      const threeColumns = mount(
        <Panel.Group>
          <Panel />
          <Panel />
          <Panel />
        </Panel.Group>,
      );
      expect(threeColumns.find('.govuk-grid-column-one-third').length).toBe(3);
      threeColumns.unmount();

      const fourColumns = mount(
        <Panel.Group>
          <Panel />
          <Panel />
          <Panel />
          <Panel />
        </Panel.Group>,
      );
      expect(fourColumns.find('.govuk-grid-column-one-quarter').length).toBe(4);
      fourColumns.unmount();

      const fourColumnsWithOther = mount(
        <Panel.Group>
          <Panel />
          <p>Test</p>
          <Panel />
          <Panel />
          <Panel />
        </Panel.Group>,
      );
      expect(fourColumnsWithOther.find('.govuk-grid-column-one-quarter').length).toBe(4);
      expect(fourColumnsWithOther.text()).toBe('Test');
      fourColumnsWithOther.unmount();

      const fiveColumns = mount(
        <Panel.Group>
          <Panel />
          <Panel />
          <Panel />
          <Panel />
          <Panel />
        </Panel.Group>,
      );
      expect(fiveColumns.find('.govuk-grid-column-full').length).toBe(5);
      fiveColumns.unmount();
    });
  });
});
