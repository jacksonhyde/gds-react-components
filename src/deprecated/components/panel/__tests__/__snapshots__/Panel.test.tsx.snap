// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel matches snapshot: Panel 1`] = `
<Panel>
  <BasePanel>
    <div
      className="govuk-panel"
    />
  </BasePanel>
</Panel>
`;
